### ISSUE TITLE
XLNet 在 DRCD dev set 的表現與 DRCD test set 的表現差異甚大

### ISSUE THREAD

**Issue by allenyummy** [ymcui/Chinese-XLNet#15]
- 問題: 使用DRCD train set 去 fine tune XLNet-Chinese，並套用此模型到DRCD dev set 與 test set，結果大不同。不知道我哪個環節出錯了..

 Model | Max sequence   Length | Batch size | Learning rate | Train steps | Dev set (EM/F1) | Test set (EM/F1)
-- | -- | -- | -- | -- | -- | --
XLNet-Chinese | 256 | 2 | 3e-5 | 12000 | 85.44 / 93.32 | 4.85 / 0.43

- 操作步驟:
1. 首先先執行 preprocess.sh，先設max seq length = 256，超過這數字，我的GPU負荷不了。
```
#!/bin/bash
#### local path
DRCD_DIR=raw_data/
INIT_CKPT_DIR=XLNet/xlnet_pretrain_model/chinese_xlnet_mid_L-24_H-768_A-12
#### google storage path
GS_ROOT=
GS_PROC_DATA_DIR=XLNet/proc_data
python3 XLNet/xlnet/run_squad.py \
  --use_tpu=False \
  --do_prepro=True \
  --spiece_model_file=${INIT_CKPT_DIR}/spiece.model \
  --train_file=${DRCD_DIR}/DRCD_training.json \
  --output_dir=${GS_PROC_DATA_DIR} \
  --uncased=False \
  --max_seq_length=256 \
  $@
```
2. Fine tune on DRCD train set，並predict dev set，最後使用 cmrc2018_evaluate.py 產生dev set 的EM和F1。

Fine tune on train set and predict dev set
```
#!/bin/bash
#### local path
DRCD_DIR=raw_data/
INIT_CKPT_DIR=XLNet/xlnet_pretrain_model/chinese_xlnet_mid_L-24_H-768_A-12
PROC_DATA_DIR=XLNet/proc_data
MODEL_DIR=XLNet/experiment/chinese_xlnet_mid_L-24_H-768_A-12_S-256_B-2
CUDA_VISIBLE_DEVICES=0,1 python3 XLNet/xlnet/run_squad.py \
  --use_tpu=False \
  --num_hosts=1 \
  --num_core_per_host=3 \
  --model_config_path=${INIT_CKPT_DIR}/xlnet_config.json \
  --spiece_model_file=${INIT_CKPT_DIR}/spiece.model \
  --output_dir=${PROC_DATA_DIR} \
  --init_checkpoint=${INIT_CKPT_DIR}/xlnet_model.ckpt \
  --model_dir=${MODEL_DIR}/model_ckpt \
  --train_file=${DRCD_DIR}/DRCD_training.json \
  --predict_file=${DRCD_DIR}/DRCD_dev.json \
  --predict_dir=${MODEL_DIR}/predict_result/dev \
  --uncased=False \
  --max_seq_length=256 \
  --do_train=True \
  --train_batch_size=2 \
  --do_predict=True \
  --predict_batch_size=32 \
  --learning_rate=3e-5 \
  --adam_epsilon=1e-6 \
  --iterations=1000 \
  --save_steps=1000 \
  --train_steps=12000 \
  --warmup_steps=1000 \
  $@
```

Evaluate on dev set 
```
#!/bin/bash
####local path
DRCD_DIR=raw_data/
EVALUATE_DIR=XLNet/xlnet/
PREDICT_RESULT=XLNet/experiment/chinese_xlnet_mid_L-24_H-768_A-12_S-256_B-2/predict_result
python2 $EVALUATE_DIR/cmrc2018_evaluate.py $DRCD_DIR/DRCD_dev.json $PREDICT_RESULT/dev/predictions.json
```
Performance on dev set
![image](https://user-images.githubusercontent.com/36063123/63565256-d3c43f00-c59a-11e9-998a-0d394b92e4a8.png)

3.  使用訓練好的XLNet，predict test set，並使用 cmrc2018_evaluate.py 產生test set 的EM和F1。

Predict test set
```
#!/bin/bash
#### local path
DRCD_DIR=raw_data/
INIT_CKPT_DIR=XLNet/xlnet_pretrain_model/chinese_xlnet_mid_L-24_H-768_A-12
INIT_CKPT_DIR_1=XLNet/experiment/chinese_xlnet_mid_L-24_H-768_A-12_S-256_B-2
PROC_DATA_DIR=XLNet/proc_data
MODEL_DIR=XLNet/experiment/chinese_xlnet_mid_L-24_H-768_A-12_S-256_B-2
CUDA_VISIBLE_DEVICES=0,1 python3 XLNet/xlnet/run_squad.py \
  --use_tpu=False \
  --num_hosts=1 \
  --num_core_per_host=3 \
  --model_config_path=${INIT_CKPT_DIR}/xlnet_config.json \
  --spiece_model_file=${INIT_CKPT_DIR}/spiece.model \
  --output_dir=${PROC_DATA_DIR} \
  --init_checkpoint=${INIT_CKPT_DIR_1}/model.ckpt-12000 \
  --model_dir=${MODEL_DIR}/model_ckpt \
  --train_file=${DRCD_DIR}/DRCD_training.json \
  --predict_file=${DRCD_DIR}/DRCD_test.json \
  --predict_dir=${MODEL_DIR}/predict_result/test \
  --uncased=False \
  --max_seq_length=256 \
  --do_train=False \
  --train_batch_size=2 \
  --do_predict=True \
  --predict_batch_size=32 \
  --learning_rate=3e-5 \
  --adam_epsilon=1e-6 \
  --iterations=1000 \
  --save_steps=1000 \
  --train_steps=12000 \
  --warmup_steps=1000 \
  $@
```

Evaluate on test set 
```
#!/bin/bash
####local path
DRCD_DIR=raw_data/
EVALUATE_DIR=XLNet/xlnet/
PREDICT_RESULT=XLNet/experiment/chinese_xlnet_mid_L-24_H-768_A-12_S-256_B-2/predict_result
python2 $EVALUATE_DIR/cmrc2018_evaluate.py $DRCD_DIR/DRCD_test.json $PREDICT_RESULT/test/predictions.json
```

Performance on test set
![image](https://user-images.githubusercontent.com/36063123/63565490-ce1b2900-c59b-11e9-93c2-b233666916f7.png)

**Comment by ymcui** [comment]
凭直觉看应该是在预测DRCD_test.json的时候网络没有被initialize。
`init_checkpoint`在单独做predict的时候是不起作用的。
接下来：
1. 你需要查看一下`${MODEL_DIR}/model_ckpt`目录下是否有xlnet的model文件。
2. 查看log中是否包含类似`Restoring parameters from`这样的信息，如果没有就说明model没有被正确加载进去。

**Comment by allenyummy** [comment]
是因為這個緣故嗎?!  感謝您的回覆!
![image](https://user-images.githubusercontent.com/36063123/63567120-86e46680-c5a2-11e9-8ce7-ea29bbb6fdac.png)

**Comment by allenyummy** [comment]
目前 `${MODEL_DIR}/model_ckpt`目錄底下的文件
![image](https://user-images.githubusercontent.com/36063123/63567178-d32fa680-c5a2-11e9-8c68-747d2ef80b51.png)

請問該如何解決model没有被正确加載進去的問題..

**Comment by ymcui** [comment]
上面那个Estimator的问题可以忽略。你需要再看一下log中是否包含`Restoring parameters from`这样的信息。

**Comment by allenyummy** [comment]
有出現這個訊息
![image](https://user-images.githubusercontent.com/36063123/63567392-fad33e80-c5a3-11e9-8b2d-ec95059b6201.png)

**Comment by ymcui** [comment]
看似weights应该是读进去了，方便的话可以把log和prediction.json发给我看看， me@ymcui.com

**Comment by allenyummy** [comment]
已寄出，十分感謝協助!!

**Comment by ymcui** [comment]
由于命名原因，dev和test生成的tf_records/features.pkl都会命名成`XLNet/proc_data/spiece.model.slen-256.qlen-64.eval.features.pkl`。
你能否改变一下`--output_dir=`到其他的目录，重新预测一遍DRCD_test.json，这样可以重新生成features和tf_records，看看会有什么样的结果。

**Comment by chiangyulun0914** [comment]
好的，我試試看，十分感謝提供建議!!!

江侑倫

國立臺灣大學生物產業機電工程學系 碩士班

Yu-Lun Chiang

Department of Bio-Industrial Mechatronics Engineering

National Taiwan University
Mobile: +886-975279013 (Taiwan)


Yiming Cui <notifications@github.com> 於 2019年8月23日 週五 下午1:53寫道：

**Comment by allenyummy** [comment]
成功了!
test set 
EM: 84.42
F1: 92.85
十分感謝!!
